scorecard/profile_data.py

```diff
@@ -236,19 +241,79 @@ def revenue_breakdown(self):
                         'item': item,
                         'amount': amount,
                         'percent': percent(amount, total),
-                        'year': year_name,
+                        'date': year_name,
                     })
                 if total and subtotal and (total != subtotal):
                     values.append({
                         'item': 'Other',
                         'amount': total - subtotal,
                         'percent': percent(total - subtotal, total),
-                        'year': year_name,
+                        'date': year_name,
                     })
             except KeyError:
                 continue
         return {'values': values}
 
+    def current_ratio(self):
+        values = []
+        results = self.results['in_year_bsheet']
+        year_key = lambda r: r['financial_year_end.year']
+        month_key = lambda r: r['financial_period.period']
+        year_sorted = sorted(results, key=year_key)
+        year_sorted.reverse()
+        quarters = {}
+        latest_quarter = None
+        for year, yeargroup in groupby(year_sorted, year_key):
+            month_sorted =  sorted(yeargroup, key=month_key)
+            month_sorted.reverse()
+            # Loop over months that exist and add their values to quarters
+            for month, monthgroup in groupby(month_sorted, key=month_key):
+                monthitems = list(monthgroup)
+                quarter_idx = ((month - 1) / 3) + 1
+                quarter_key = (year, quarter_idx)
+                try:
+                    # Rely on index out of range for missing values to skip month if one's missing
+                    assets = [m['amount.sum'] for m in monthitems if m['item.code'] == '2150'][0]
+                    liabilities = [m['amount.sum'] for m in monthitems if m['item.code'] == '1600'][0]
+
+                    if quarter_key not in quarters:
+                        q = {
+                            'date': "%sq%s" % quarter_key,
+                            'year': year,
+                            'month': month,
+                            'amount_type': 'ACT',
+                            'quarter': quarter_idx,
+                            'assets': assets,
+                            'liabilities': liabilities,
+                            'result': ratio(assets, liabilities),
+                        }
+                        quarters[quarter_key] = q
+                        if latest_quarter is None:
+                            latest_quarter = q
+                except IndexError:
+                    pass
+        # Enumerate the quarter keys we can expect to exist based on the latest
+        keys = []
+        result_quarters = []
+        if latest_quarter is not None:
```

Comment:
I like how short it is, however for generating the keys, I'm hesitant because it takes me a long time to parse what's actually happening there. I'd rather spend 5 lines for my version that's closer to braindead simple than one that's very easy to break and very hard to debug.

I like your buildup of values array though - so my key generation plus your `values` generation with an array slice (since I generate too many keys) should be a good compromise?